### ISSUE TITLE
Hello , If I want to detect an custom moving object in  addition with coco dataset which files need to be customized?

### ISSUE THREAD

**Issue by DimaMirana** [maxmarkov/track_and_count#1]
I want to detect a person and keep tracking of that person using Deep-Sort. Now I want to detect the hands and faces of each tracked person. Which files needs to be modified?

**Comment by maxmarkov** [comment]
Hi, you should use the track_yolov5_pose.py script. It combines the yolov5 detector with Deep Sort tracker and AlphaPose pose estimator. It does a pose estimation for all detected persons (not a chosen person) on a video. By default, it uses a neural network trained on a COCO dataset with 17 points. You can also use networks trained on HALPE datasets with 26 and 136 points as well.

**Comment by DimaMirana** [comment]
Or if I want to detect car and people in the video and track only people, where needs to change. It looks like the yolov5 model here only detects the person labels and excludes the other. Now if I want to detects the labels in yolo and count and track only the person which files needs to be modified?

**Comment by PraveenGit3** [comment]
Hi Why does track works on only person . If i used my custom  Yolov5 weights which i used for parking slot ? Does it track all bounding boxes ? or only for persons?

**Comment by maxmarkov** [comment]
> Hi Why does track works on only person . If i used my custom Yolov5 weights which i used for parking slot ? Does it track all bounding boxes ? or only for persons?

Hi, 

There are 2 different trackers used in this repository. One is based on the motion model (it is called '**sort**'), and another one which is based on the appearance model (it is called '**DeepSort**'). The former model evaluates the state vector (position+velocity) for each object (of any type) fed into the tracker. Then, the tracker compares the state vector in a new frame with the prediction done based on the state vectors in the previous frames (the Kalman filter). The predictions and measurements of state vectors for all objects are then compared to find the correspondence between the objects in previous frames and the current one (the Hungarian algorithm). One can use the sort tracker for any type of object since it requires the object's position and velocity for tracking, but not the information specific to the object's type (a person, a car, etc). 

On the contrary, the DeepSort algorithm uses a neural network trained to distinguish different people based on their appearance. When Yolo identifies a person, its cropped image (i.e what is inside the bounding box) is fed to the neural network to get a feature vector representing this person. Then, you can track people by calculating the difference between these feature vectors with some metric (cosine metric in this case). Note that the neural network has been trained uniquely on people and, thus, passing the custom object to it would be erroneous. If you want to use the DeepSort for your custom object, you will need to retrain the neural network inside the DeepSort on your custom object dataset. 

In my experience, the sort algorithm works as well as the DeepSort. You just need to experiment with the parameters a little bit.
Check the max_age parameter responsible for the maximum number of frames to keep alive a track without associated detections (sometimes the object is lost, but needs to be kept in memory to restore the tracking).

If you want to use the sort algorithm, run the following script
python3 track_yolov5_sort.py --source YOUR_VIDEO.mp4 --weights yolov5/weights/yolov5s.pt --conf 0.4 --max_age 50 --min_hits 10 --iou_threshold 0.3 

There is also one undocumented option. You can run the DeepSort tracker for a 'person' class (let's say with the default pre-trained model like yolov5s.pt), and the Sort tracker for your custom object class with your custom model:

python3 track_yolov5_ensemble.py --source YOUR_VIDEO.mp4 --weights yolov5/weights/yolov5s.pt YOUR_CUSTOM_MODEL.pt --max_age 50 --min_hits 10 --iou_threshold 0.3

You just need to change the ballot class to your custom object in track_yolov5_ensemble.py   

Hope it helps.
